### PR TITLE
fixed #3187

### LINKED ISSUE #3187: TransformNode.setParent does not preserve position, rotation, scaling

### PR DESCRIPTION
**Pull request by abow**

### CODE REVIEW

File: src/Mesh/babylon.transformNode.ts [hunk 1]
```diff
@@ -451,14 +454,43 @@ module BABYLON {
...
+                
+                parent.computeWorldMatrix(true);
+                parent.getWorldMatrix().decompose(scale, rotation, position);
+                
+                rotation.toRotationMatrix(m1);
+                m2.setTranslation(position);
+                
+                m2.multiplyToRef(m1, m1);
+                
+                var invParentMatrix = Matrix.Invert(m1);
```
Comment (deltakosh) [inline src/Mesh/babylon.transformNode.ts]: perhaps we could rely on ToRef versions to avoid GC ?

**Comment by deltakosh** [comment]
Also could you change the signature to:
public setParent(node: Nullable<Node>): TransformNode {

**Comment by abow** [comment]
No problem.  I just came up with another test and found an error.

File: src/Mesh/babylon.transformNode.ts [hunk 2]
```diff
@@ -448,18 +450,44 @@ module BABYLON {
...
+                
+                node.computeWorldMatrix(true);
+                node.getWorldMatrix().decompose(scale, rotation, position);
+                
+                rotation.toRotationMatrix(m1);
+                m2.setTranslation(position);   
+                m2.multiplyToRef(m1, m1);     
+                m1.invertToRef(invParentMatrix);
+                
+                var m = this.getWorldMatrix().multiply(invParentMatrix);
```
Comment (deltakosh) [inline src/Mesh/babylon.transformNode.ts]: Perhaps the multiply can be saved as well

**Comment by deltakosh** [comment]
Thanks mate!

**Pull request merged**